chia/full_node/mempool_manager.py

```diff
@@ -286,42 +285,94 @@ async def pre_validate_spendbundle(
+        The mempool should be locked during this call (blockchain lock). If there are mempool conflicts, the conflicting
+        spends might be removed (if the new spend is a superset of the previous). Otherwise, the new spend might be
+        added to the potential pool.
+
+        Args:
+            new_spend: spend bundle to validate and add
+            npc_result: result of running the clvm transaction in a fake block
+            spend_name: hash of the spend bundle data, passed in as an optimization
+
+        Returns:
+            Optional[uint64]: cost of the entire transaction, None iff status is FAILED
+            MempoolInclusionStatus:  SUCCESS (should add to pool), FAILED (cannot add), and PENDING (can add later)
+            Optional[Err]: None iff status is FAILED
+        """
+
+        # Skip if already added
+        if spend_name in self.mempool.spends:
+            cost: Optional[uint64] = self.mempool.spends[spend_name].cost
+            assert cost is not None
+            return uint64(cost), MempoolInclusionStatus.SUCCESS, None
+
+        cost, status, err, item, remove_items = await self.validate_spend_bundle(new_spend, npc_result, spend_name)
+        if status == MempoolInclusionStatus.FAILED:
+            assert cost is None and item is None
+        elif status == MempoolInclusionStatus.SUCCESS:
+            assert cost is not None and item is not None
+            self.mempool.remove_from_pool(remove_items)
+            self.mempool.add_to_pool(item)
+        elif status == MempoolInclusionStatus.PENDING:
+            assert cost is not None and item is not None
+            self.potential_cache.add(item)
+        else:
+            raise RuntimeError(f"Did not expect mempool status: {status}")
+        return cost, status, err
+
+    async def validate_spend_bundle(
+        self,
+        new_spend: SpendBundle,
+        npc_result: NPCResult,
+        spend_name: bytes32,
+    ) -> Tuple[Optional[uint64], MempoolInclusionStatus, Optional[Err], Optional[MempoolItem], List[bytes32]]:
+        """
+        Validates new_spend with the given NPCResult, and spend_name, and the current mempool. The mempool should
+        be locked during this call (blockchain lock).
+
+        Args:
+            new_spend: spend bundle to validate
+            npc_result: result of running the clvm transaction in a fake block
+            spend_name: hash of the spend bundle data, passed in as an optimization
+
+        Returns:
+            Optional[uint64]: cost of the entire transaction, None iff status is FAILED
+            MempoolInclusionStatus:  SUCCESS (should add to pool), FAILED (cannot add), and PENDING (can add later)
+            Optional[Err]: None iff status is FAILED
+            Optional[MempoolItem]: the item to add (None iff status is FAILED)
+            List[bytes32]: conflicting mempool items to remove empty unless status is SUCCESS
```

Comment:
I think you meant to add a "." between "remove" and "empty". It took a minute to parse :)